lib/c.c

```diff
@@ -352,71 +355,27 @@ void printf(char *str, ...)
             si++;
         }
     }
-    buffer[bi] = 0;
-    __syscall(__syscall_write, 1, buffer, bi);
+
+    int len = size - 1 > bi ? bi : size - 1;
+    buffer[len] = 0;
+    return len;
 }
 
-void sprintf(char *buffer, char *str, ...)
+int printf(char *str, ...)
 {
-    int *var_args = &str + 4;
-    int si = 0, bi = 0, pi = 0;
+    char buffer[200];
+    int len = __format_to_buf(buffer, str, &str + 4, 0x7fffffff);
+    return __syscall(__syscall_write, 1, buffer, len);
+}
 
-    while (str[si]) {
-        if (str[si] != '%') {
-            buffer[bi] = str[si];
-            bi++;
-            si++;
-        } else {
-            int w = 0, zp = 0, pp = 0;
+int sprintf(char *buffer, char *str, ...)
+{
+    return __format_to_buf(buffer, str, &str + 4, 0x7fffffff);
```

Comment:
`0x7fffffff` looks tricky.

Reply:
Perhap introducing `INT_MAX` and `INT_MIN` originally defined from `limits.h` to replace these obscure value might be better?